### PR TITLE
[shopsys] update of several packages and different way to install Phing

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, the reason for the PR| We want to keep our dependencies updated and this has been neglected for some time, so this is the first step towards the newest versions. 
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

- This PR adds a new PHPUnit 11 (upgrade from 9) that introduced many changes, so the tests have many updates. 
- Because of decreased support for extensions, `zalas/phpunit-injector` cannot be installed the same way as before, and currently, there is a hybrid installation of the latest dev version of this package, as it is the only version, where PHPUnit 11 is supported + our workaround to inject service in the `WebTestCase`. see https://github.com/jakzal/phpunit-injector/issues/33 for more information.
- Phing is now downloaded and installed as PHAR file. This removes one dependency that needed to include all the latest versions of packages we wanted to upgrade to and often blocked us.

----
:globe_with_meridians: Live Preview:
  - https://tl-update-composer-dependencies.odin.shopsys.cloud
  - https://cz.tl-update-composer-dependencies.odin.shopsys.cloud

### CODE REVIEW

File: packages/article-feed-luigis-box/phpunit.xml [hunk 1]
```diff
@@ -2,16 +2,14 @@
 <!-- http://www.phpunit.de/manual/current/en/appendixes.configuration.html -->
 <phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     backupGlobals="false"
-    backupStaticAttributes="false"
     colors="true"
-    convertErrorsToExceptions="true"
-    convertNoticesToExceptions="true"
-    convertWarningsToExceptions="true"
```
Comment (vitek-rostislav) [inline packages/article-feed-luigis-box/phpunit.xml:9]: are these settings taken into account now? From what I googled, it should be written in the setup method of the test case, do we do that now?

File: packages/article-feed-luigis-box/tests/Unit/LuigisBoxArticleFeedItemTest.php [hunk 2]
```diff
@@ -47,7 +48,6 @@ public function articleFeedItemCreationDataProvider(): iterable
                 ...$commonArticleData,
                 'imageUrl' => self::ARTICLE_IMAGE_URL,
             ],
-            'expectedImageUrl' => self::ARTICLE_IMAGE_URL . '?width=605',
```
Comment (vitek-rostislav) [inline packages/article-feed-luigis-box/tests/Unit/LuigisBoxArticleFeedItemTest.php:50]: would not it be better to add an assert for the image URL?

File: packages/article-feed-luigis-box/phpunit.xml (hunk 1 above)
Comment (vitek-rostislav) [reply packages/article-feed-luigis-box/phpunit.xml:9]: the same question applies to all the config files

File: packages/coding-standards/tests/Unit/PhpStan/GedmoExtensionTest.php [hunk 3]
```diff
@@ -40,10 +42,11 @@ protected function setUp(): void
     }
 
     /**
-     * @dataProvider getProperties
      * @param string $propertyName
      * @param bool $isWritten
      */
+    #[DataProvider('getProperties')]
+    #[RunInSeparateProcess]
```
Comment (vitek-rostislav) [inline packages/coding-standards/tests/Unit/PhpStan/GedmoExtensionTest.php:49]: why? :slightly_smiling_face:

File: packages/framework/tests/Test/MoneyExporter.php [hunk 4]
```diff
@@ -4,23 +4,14 @@
...
      */
-    protected function recursiveExport(&$value, $indentation, $processed = null): string
+    public function export(mixed $value): string
     {
-        if ($value instanceof Money) {
-            return $value->getAmount();
-        }
-
-        return parent::recursiveExport($value, $indentation, $processed);
+        return $value->getAmount();
```
Comment (vitek-rostislav) [inline packages/framework/tests/Test/MoneyExporter.php:17]: this would fail when `$value` is not an instance of `Money`, or? Why is it like this? Maybe we can typehint the parameter with `Money` type?

File: packages/framework/tests/Test/PriceExporter.php [hunk 5]
```diff
@@ -4,23 +4,14 @@
...
      */
-    protected function recursiveExport(&$value, $indentation, $processed = null): string
+    public function export(mixed $value): string
     {
-        if ($value instanceof Price) {
-            return $value->getPriceWithVat()->getAmount() . ' with Vat (' . $value->getPriceWithoutVat()->getAmount() . ' without VAT)';
-        }
-
-        return parent::recursiveExport($value, $indentation, $processed);
+        return $value->getPriceWithVat()->getAmount() . ' with Vat (' . $value->getPriceWithoutVat()->getAmount() . ' without VAT)';
```
Comment (vitek-rostislav) [inline packages/framework/tests/Test/PriceExporter.php:17]: same as above, $value needs to be an instance of Price, otherwise it might fail

File: packages/framework/tests/Unit/Command/CheckRedisCommandTest.php [hunk 6]
```diff
@@ -17,30 +18,33 @@ final class CheckRedisCommandTest extends TestCase
...
+     * @param array $mockMethodNames
      */
-    public function testPingAllRedisClients(bool $expectSuccess, RedisFacade $redisFacade): void
+    #[DataProvider('pingAllRedisClientsProvider')]
+    public function testPingAllRedisClients(bool $expectSuccess, array $mockMethodNames): void
     {
+        $redisMocks = [];
+
+        foreach ($mockMethodNames as $mockMethodName) {
+            $redisMocks[] = $this->{$mockMethodName}();
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Command/CheckRedisCommandTest.php:44]: nice :+1:

File: packages/framework/tests/Unit/Component/Translation/PhpFileExtractorTest.php [hunk 7]
```diff
@@ -67,7 +66,8 @@ private function extract($filename)
         $extractor = $this->getExtractor();
 
         $lexer = new Lexer();
-        $parser = new Multiple([new Php7($lexer), new Php5($lexer)]);
+        $parserFactory = new ParserFactory();
+        $parser = $parserFactory->createForVersion(PhpVersion::fromString('8.3'));
```
Comment (vitek-rostislav) [inline packages/framework/tests/Unit/Component/Translation/PhpFileExtractorTest.php:70]: just an idea, could we add a comment somewhere to dockerfile or composer json so we know we should update these occurrences when upgrading PHP version?

File: packages/frontend-api/tests/Unit/Model/Token/TokenFacadeTest.php [hunk 8]
```diff
@@ -27,12 +28,13 @@
...
-     * @dataProvider tokensDataProvider
-     * @param \Lcobucci\JWT\UnencryptedToken $token
+     * @param array $tokenData
      * @param class-string|null $exceptionClass
      */
-    public function testTokenValidation(UnencryptedToken $token, ?string $exceptionClass): void
+    #[DataProvider('tokensDataProvider')]
+    public function testTokenValidation(array $tokenData, ?string $exceptionClass): void
     {
+        $token = $this->createToken($tokenData['issuedBy'], $tokenData['privateKey'], $tokenData['expiresAt']);
```
Comment (vitek-rostislav) [inline packages/frontend-api/tests/Unit/Model/Token/TokenFacadeTest.php]: it would be nice to have these keys in class constants

File: packages/migrations/tests/Unit/Component/Doctrine/Migrations/AbstractMigrationTest.php [hunk 9]
```diff
@@ -1,29 +0,0 @@
...
-declare(strict_types=1);
-
-namespace Tests\MigrationBundle\Unit\Component\Doctrine\Migrations;
-
-use PHPUnit\Framework\TestCase;
-use ReflectionClass;
-use Shopsys\MigrationBundle\Component\Doctrine\Migrations\AbstractMigration;
-use Shopsys\MigrationBundle\Component\Doctrine\Migrations\Exception\MethodIsNotAllowedException;
-
-class AbstractMigrationTest extends TestCase
```
Comment (vitek-rostislav) [inline packages/migrations/tests/Unit/Component/Doctrine/Migrations/AbstractMigrationTest.php]: why is this removed?

File: UPGRADE-15.0.md [hunk 10]
```diff
@@ -1050,6 +1050,16 @@ Follow the instructions in relevant sections, e.g. `shopsys/coding-standards` or
     -   `NavigationQuery`
 -   see #project-base-diff to update your project
 
+#### update composer dependencies to newer versions
```
Comment (vitek-rostislav) [inline UPGRADE-15.0.md]: link to the PR is missing

File: packages/framework/src/Twig/NumberFormatterExtension.php [hunk 11]
```diff
@@ -54,11 +54,11 @@ public function getFilters()
     }
 
     /**
-     * @param mixed $number
+     * @param string $number
      * @param string|null $locale
      * @return string
      */
-    public function formatNumber($number, $locale = null)
+    public function formatNumber(string $number, ?string $locale = null): string
```
Comment (vitek-rostislav) [inline packages/framework/src/Twig/NumberFormatterExtension.php:61]: is it ok that you are now forced to use string only? I thought that these filters could be used with various types (float, int, string)?

File: phing [hunk 12]
```diff
@@ -1,11 +1,7 @@
 #!/usr/bin/env php
 <?php
 
-/**
- * Shell wrapper for Phing
- * $Id: 3da5a2758ca689a6ee49defabcda5efa6dd70a97 $
- */
-$phingPhpFilepath = __DIR__ . '/vendor/phing/phing/bin/phing.php';
+$phingPhpFilepath = __DIR__ . '/phing.phar';
```
Comment (vitek-rostislav) [inline phing]: nitpick, the variable should be rather named `$phingPharFilepath` now

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]
Hi, really nice job there, I appreciate the new version of PHPUnit, as well as the phing as phar :ok_hand:  I have added a few notes, plus please check the filing build, there is a problem in dumping translations - `dir` option is unknown now

File: packages/article-feed-luigis-box/phpunit.xml (hunk 1 above)
Comment (TomasLudvik) [reply packages/article-feed-luigis-box/phpunit.xml:9]: We decided to keep it as it is

File: packages/coding-standards/tests/Unit/PhpStan/GedmoExtensionTest.php (hunk 3 above)
Comment (TomasLudvik) [reply packages/coding-standards/tests/Unit/PhpStan/GedmoExtensionTest.php:49]: Because PHPStan/PHPUnit uses old version of PHPUnit and it was causing problems for other tests.

File: packages/framework/tests/Unit/Component/Translation/PhpFileExtractorTest.php (hunk 7 above)
Comment (TomasLudvik) [reply packages/framework/tests/Unit/Component/Translation/PhpFileExtractorTest.php:70]: I think that this new format now has higher chance for future upgrades to be found and updated as well :)

File: packages/frontend-api/tests/Unit/Model/Token/TokenFacadeTest.php (hunk 8 above)
Comment (TomasLudvik) [reply packages/frontend-api/tests/Unit/Model/Token/TokenFacadeTest.php]: I have expanded it to separate method parameters

File: packages/framework/src/Twig/NumberFormatterExtension.php (hunk 11 above)
Comment (TomasLudvik) [reply packages/framework/src/Twig/NumberFormatterExtension.php:61]: We decided to keep it as it is because it is better to let developer to parse number correctly to string and sent it here.

File: phing (hunk 12 above)
Comment (TomasLudvik) [reply phing]: I have renamed it

File: packages/migrations/tests/Unit/Component/Doctrine/Migrations/AbstractMigrationTest.php (hunk 9 above)
Comment (TomasLudvik) [reply packages/migrations/tests/Unit/Component/Doctrine/Migrations/AbstractMigrationTest.php]: I have restored it

File: packages/article-feed-luigis-box/tests/Unit/LuigisBoxArticleFeedItemTest.php (hunk 2 above)
Comment (TomasLudvik) [reply packages/article-feed-luigis-box/tests/Unit/LuigisBoxArticleFeedItemTest.php:50]: I have updated the test

**Review by vitek-rostislav — APPROVED** [review]

File: composer.json [hunk 13]
```diff
@@ -221,7 +221,7 @@
         ],
         "security-check": "security-checker security:check",
         "auto-scripts": {
-            "shopsys:phing:download": "symfony-cmd",
+            "php project-base/app/app/downloadPhing.php": "script",
```
Comment (vitek-rostislav) [inline composer.json:225]: shopsys:phing:download is still mentioned in the upgrade notes

File: packages/framework/src/Component/Phing/PhingDownloader.php [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+namespace Shopsys\FrameworkBundle\Component\Phing;
+
+use PharIo\Version\InvalidVersionException;
+use PharIo\Version\Version;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class PhingDownloader
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Phing/PhingDownloader.php:14]: `DownloadPhingCommand` should be now removed, or?

File: packages/framework/src/Component/Phing/PhingDownloader.php [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+{
+    public const string OPTION_VERSION = 'phing-version';
+    protected const string VERSION_URL = 'https://github.com/phingofficial/phing/releases/download/%s/phing-%s.phar';
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Phing/PhingDownloader.php:17]: I know we were already discussing it but would not it be easier to use just `https://www.phing.info/get/phing-latest.phar`? You would not have to do the magic with the version resolving

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]

File: packages/framework/src/Component/Phing/PhingDownloader.php (hunk 15 above)
Comment (TomasLudvik) [reply packages/framework/src/Component/Phing/PhingDownloader.php:17]: Nope as we want projects to be on locked versions to ensure compatibility.

File: upgrade-notes/backend_20240529_084428.md [hunk 16]
```diff
@@ -1017,7 +1017,7 @@
     -   see #project-base diff to see changes you might need to apply in your tests
 -   `commerceguys/intl` has been updated to the latest version
     -   `IntlCurrencyRepository` and `NumberFormatterExtension` class methods have updated their interfaces to include strict types, you will need to update your usages of such methods in your project
-    -   `Phing` is no longer direct dependency, but is downloaded separately via `shopsys:phing:download` command. This is done directly in `composer.json` file, so you should follow changes in project-base to update `composer.json`, `phing` and other files accordingly
+    -   `Phing` is no longer direct dependency, but is downloaded separately via `php project-base/app/app/downloadPhing.php` command. This is done directly in `composer.json` file, so you should follow changes in project-base to update `composer.json`, `phing` and other files accordingly
```
Comment (vitek-rostislav) [inline upgrade-notes/backend_20240529_084428.md:1020]: I think now, with the new approach, in the upgrading file, there should be your pull request only, am I right?

**Review by vitek-rostislav — APPROVED** [review]
